File: frontend/src/admin/products/Product.ts
import { z } from 'zod'

export const Kategorie = {
  ESSEN: 'essen',
  GETRAENK: 'getraenk',
  SONSTIGES: 'sonstiges',
} as const
export type Kategorie = (typeof Kategorie)[keyof typeof Kategorie]

export const VarianteStatus = {
  ACTIVE: 'active',
  INACTIVE: 'inactive',
} as const
export type VarianteStatus =
  (typeof VarianteStatus)[keyof typeof VarianteStatus]

export const ProduktIdSchema = z.number().int().min(1)
export const VarianteIdSchema = z.number().int().min(1)

const NameSchema = z
  .string()
  .min(3, { message: 'Das sieht nicht nach einem echten Namen aus.' })
  .max(100, { message: 'Der Name ist zu lang.' })
const PreisCentsSchema = z
  .number()
  .int()
  .min(0, { message: 'Preis muss mindestens 0 Cent sein.' })
const KategorieSchema = z.enum(['essen', 'getraenk', 'sonstiges'])
const VarianteStatusSchema = z.enum(['active', 'inactive'])
const DateStringSchema = z.string().refine((date) => !isNaN(Date.parse(date)), {
  message: 'Ungültiges Datumsformat',
})

export const VarianteSchema = z.object({
  id: VarianteIdSchema,
  name: NameSchema,
  preisCents: PreisCentsSchema,
  status: VarianteStatusSchema,
  createdAt: DateStringSchema,
})
export type Variante = z.infer<typeof VarianteSchema>

export const ProduktSchema = z.object({
  id: ProduktIdSchema,
  name: NameSchema,
  kategorie: KategorieSchema,
  varianten: z.array(VarianteSchema),
  createdAt: DateStringSchema,
})
export type Produkt = z.infer<typeof ProduktSchema>

File: backend/domain/produkt/product.go
package produkt

import (
	"fmt"
	"math"
	"time"

	z "github.com/Oudwins/zog"
	"github.com/nicograef/jotti/backend/domain/steuer"
)

type Kategorie string

const (
	EssenKategorie     Kategorie = "essen"
	GetraenkKategorie  Kategorie = "getraenk"
	SonstigesKategorie Kategorie = "sonstiges"
)

// Richtung ist die Verschieberichtung in der Anzeigereihenfolge. Die Reihenfolge
// selbst bleibt reine Persistenz: kein Feld am Aggregat, keine Response liefert
// sie — das Backend gibt die fertig sortierte Liste aus.
type Richtung string

const (
	RichtungHoch   Richtung = "hoch"
	RichtungRunter Richtung = "runter"
)

type Produkt struct {
	ID         int
	Name       string
	Kategorie  Kategorie
	Steuersatz steuer.Steuersatz
	Status     Status
	Varianten  []Variante
	CreatedAt  time.Time
	UpdatedAt  time.Time
}

// IDSchema is bounded at both ends. The upper bound is the largest value the
// int4 column holds: a request carrying more can only be wrong, and the schema
// answers 400 instead of letting it fail inside the driver.
var IDSchema = z.Int().
	GTE(1, z.Message("Ungültige Produkt-ID")).
	LTE(math.MaxInt32, z.Message("Ungültige Produkt-ID"))

var NameSchema = z.String().Trim().Min(3, z.Message("Name zu kurz")).Max(100, z.Message("Name zu lang"))

var KategorieSchema = z.StringLike[Kategorie]().OneOf(
	[]Kategorie{EssenKategorie, GetraenkKategorie, SonstigesKategorie},
	z.Message("Ungültige Kategorie"),
)

var RichtungSchema = z.StringLike[Richtung]().OneOf(
	[]Richtung{RichtungHoch, RichtungRunter},
	z.Message("Ungültige Richtung"),
)

var SteuersatzSchema = steuer.SteuersatzSchema

var ProduktSchema = z.Struct(z.Shape{
	"ID":         IDSchema.Required(),
	"Name":       NameSchema.Required(),
	"Kategorie":  KategorieSchema.Required(),
	"Steuersatz": SteuersatzSchema.Required(),
	"Status":     StatusSchema.Required(),
	"Varianten":  z.Slice(VarianteSchema).Required(),
	"CreatedAt":  z.Time().Required(),
	"UpdatedAt":  z.Time().Required(),
})

func (p Produkt) Validate() error {
	if errs := ProduktSchema.Validate(&p); errs != nil {
		issues := z.Issues.FlattenAndCollect(errs)
		return fmt.Errorf("invalid product: %v", issues)
	}
	return nil
}

// NewProdukt assigns no ID; the persistence layer sets it.
func NewProdukt(name string, kategorie Kategorie, steuersatz steuer.Steuersatz) (Produkt, error) {
	if issue := NameSchema.Validate(&name); issue != nil {
		return Produkt{}, fmt.Errorf("invalid name")
	}

	if issue := KategorieSchema.Validate(&kategorie); issue != nil {
		return Produkt{}, fmt.Errorf("invalid category")
	}

	if issue := SteuersatzSchema.Validate(&steuersatz); issue != nil {
		return Produkt{}, fmt.Errorf("invalid tax rate")
	}

	produkt := Produkt{
		Name:       name,
		Kategorie:  kategorie,
		Steuersatz: steuersatz,
		Status:     ActiveStatus,
		Varianten:  []Variante{},
		CreatedAt:  time.Now().UTC(),
		UpdatedAt:  time.Now().UTC(),
	}

	return produkt, nil
}

func (p *Produkt) UpdateDetails(name string, kategorie Kategorie, steuersatz steuer.Steuersatz) error {
	if issue := NameSchema.Validate(&name); issue != nil {
		return fmt.Errorf("invalid name")
	}

	if issue := KategorieSchema.Validate(&kategorie); issue != nil {
		return fmt.Errorf("invalid category")
	}

	if issue := SteuersatzSchema.Validate(&steuersatz); issue != nil {
		return fmt.Errorf("invalid tax rate")
	}

	p.Name = name
	p.Kategorie = kategorie
	p.Steuersatz = steuersatz
	p.UpdatedAt = time.Now().UTC()

	return nil
}

func (p *Produkt) Delete() {
	p.Status = DeletedStatus
	p.UpdatedAt = time.Now().UTC()
}
